### PR TITLE
Add browser activity filter

### LINKED ISSUE #196: Ability to filter browsing activity by url
We should allow including or excluding websites based on regex patterns. This means a toggle between `Include Only` and `Exclude`, with a textarea or input list of configurable regex patterns. The patterns would run on the whole url so they could match just the domain, path parts, or both.

We should set some sane defaults for known dev sites, similar to https://github.com/wakatime/browser-wakatime/issues/249.

### PR DESCRIPTION
**Pull request by starbugs**
I took the liberty to redesign the Settings window a bit:
Screenshot 2024-03-21 at 13 33 32
The whitelist also supports the `@@` syntax.

Closes #196

### CODE REVIEW

**Comment by alanhamlett** [comment]
I don't like the `@@` syntax, let's use structured UI and data?

**Comment by starbugs** [comment]
> I don't like the `@@` syntax, let's use structured UI and data?

Not sure I understand exactly what you mean...

**Comment by starbugs** [comment]
> Let's rename to `allow` and `deny` lists, like: `all except denied sites`

This is renamed now.

Let me know what to do about the @@ syntax then I will finalize this PR.

File: WakaTime/Helpers/PropertiesManager.swift [hunk 1]
```diff
@@ -71,4 +79,59 @@ class PropertiesManager {
...
+            UserDefaults.standard.set(newValue, forKey: Keys.denylist.rawValue)
+            UserDefaults.standard.synchronize()
+        }
+    }
+
+    static var allowlist: String {
+        get {
+            guard let allowlist = UserDefaults.standard.string(forKey: Keys.allowlist.rawValue) else {
+                return
+                    "https://github.com/\n" +
```
Comment (alanhamlett) [inline WakaTime/Helpers/PropertiesManager.swift:115]: Let's use just the domain part, not protocol nor path. So no slashes necessary.

**Comment by alanhamlett** [comment]
For `@@`, we should build into the UI the ability to assign a project to a domain. We want to build a UI similar to [custom rules](https://wakatime.com/faq#rename-projects) on the website, where we have a button `Add new project rule` that adds a new form row with two input fields `url regex` and `project`. When the full url matches a regex, we use the corresponding project name.

For now, let's just not support the `@@` syntax and build the `Custom Project Regex` feature in a future PR.

**Comment by alanhamlett** [comment]
Also, we need a setting to toggle sending `full url` or `domain-only` for browsing activity.

**Comment by starbugs** [comment]
> Actually, let's make the allow/deny list regexes which can match the full url.

I implemented that and removed the @@ syntax.

**Comment by alanhamlett** [comment]
> > Actually, let's make the allow/deny list regexes which can match the full url.
> 
> I implemented that and removed the @@ syntax.

Thank you! That's so much better than `@@` syntax.

File: WakaTime/Helpers/FilterManager.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+            // Create scheme-prefixed address versions to allow regular expressions
+            // that incorporate a scheme to match
+            let httpAddress = "http://" + address
+            let httpsAddress = "https://" + address
+
+            switch PropertiesManager.filterType {
+                case .denylist:
+                    for pattern in patterns {
+                        if httpAddress.matchesRegex(pattern) || httpsAddress.matchesRegex(pattern) {
```
Comment (alanhamlett) [inline WakaTime/Helpers/FilterManager.swift:19]: Also try matching just `address`:

```swift
if address.matchesRegex(pattern) || httpAddress.matchesRegex(pattern) || httpsAddress.matchesRegex(pattern) {
```

Comment (starbugs) [reply WakaTime/Helpers/FilterManager.swift]: Done

**Review by alanhamlett — APPROVED** [review]
🚀  Devs have waited for a long time for this feature... you've made them very happy!

**Pull request merged**